File: LeetCode/Study Plan/Algorithms/algo-2/Sliding Window/subarray_product_less_than_k.go
/*
Given an array of integers nums and an integer k, return the number of contiguous subarrays
where the product of all the elements in the subarray is strictly less than k.
*/

package main

import "fmt"

func subarray_product(arr []int, k int) int {
	if k < 1 {
		return 0
	}
	cnt := 0
	i, j := 0, 0
	n := len(arr)
	val := 1
	for i < n && j < n {
		val *= arr[j]
		for val >= k && i < n {
			val /= arr[i]
			i++
			if i > n {
				break
			}
		}
		if i < n {
			cnt += j - i + 1
		}
		j++
	}
	return cnt
}

func main() {
	fmt.Println(subarray_product([]int{10, 5, 2, 6}, 100))
	fmt.Println(subarray_product([]int{1, 2, 3}, 0))
	fmt.Println(subarray_product([]int{1, 1, 1}, 1))
}

File: LeetCode/Study Plan/Practice Problems/Medium/Subarray Prodcut Less Than K/Java/subarrayProductLessThanK.java
class Solution {
    public int numSubarrayProductLessThanK(int[] nums, int k) {
        int n = nums.length;
        int i = 0;
        int j = 0;
        int val = 1;
        int cnt = 0;
        while (i < n && j < n ) {
            val *= nums[j];
            while(val >= k && i < n) {
                val /= nums[i];
                i++;
            }
            if (i < n) cnt += j-i+1;
            j++;
        }
        return cnt;
    }
}
